### PR TITLE
Characterization tests 21-09-24-1

### PR DESCRIPTION
**Pull request by enekesabel**
#### E2E Characterization Testing

- [x] I have written at least one success scenario and at least one failure scenario for each endpoint
- [x] I have used Gherkins to write my E2E acceptance tests which characterize the API endpoints

#### Idempotency

- [x] I can run all e2e tests over and over; they work (due to idempotency & randomness using the builder), the successful setup & teardown of the database, or both

#### Builders & Fixtures

- [x] I have confirmed that all tests pass
- [x] I have used Database Fixtures and Builders to handle the edge (failure) cases

#### Expressive builders

- [x] I have written builders in declarative, expressive language which clearly expresses the state of the world (the initial state, fixture) in plain English

### Personal notes

I didn't dig that deep into improving my builders, because I think the project's current state didn't require that much complexity.
Ex.: Creating student assignments is possible even if the student is not enrolled to any class. Since we're characterizing the bugs as well, I ignored the precondition of enrollment which let me use simpler builder setup.

### CODE REVIEW

File: ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart1/exercises/1_RefactoringTo4Tiers/begin/tests/e2e/assignStudentToAssignment.test.ts [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+const feature = loadFeature(
+  path.join(__dirname, "../features/assignStudentToAssignment.feature")
+);
+
+defineFeature(feature, (test) => {
+
+    beforeEach(resetDatabase);
+
+    afterAll(() => {
+        app.close();
```
Comment (MarioBanovac) [inline ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart1/exercises/1_RefactoringTo4Tiers/begin/tests/e2e/assignStudentToAssignment.test.ts:18]: I didn't know this was possible with Express?

File: ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart1/exercises/1_RefactoringTo4Tiers/begin/tests/e2e/assignStudentToAssignment.test.ts [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+    const givenIHaveAStundentAndAnAssignment = ({given, and}:{given: DefineStepFunction, and: DefineStepFunction}) => {
+        given('I have a student', async() => {
+            student = await StudentBuilder.Fake().build();
+        });
+
+        and('I have an assignment', async () => {
+            assignment = await AssignmentBuilder.Fake().build();
+        });
+    }
```
Comment (MarioBanovac) [inline ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart1/exercises/1_RefactoringTo4Tiers/begin/tests/e2e/assignStudentToAssignment.test.ts:32]: I'm just speaking my opinion :) but I think is an unnecessary abstraction. For me it's clearer if you put it directly under the test below

File: ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart1/exercises/1_RefactoringTo4Tiers/begin/tests/e2e/createAssignment.test.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import { app, Errors } from "../../src";
+import { resetDatabase, ClassRoomBuilder } from "../fixtures";
+
+const feature = loadFeature(
+  path.join(__dirname, "../features/createAssignment.feature")
+);
+
+defineFeature(feature, (test) => {
+
+    beforeEach(resetDatabase);
```
Comment (MarioBanovac) [inline ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart1/exercises/1_RefactoringTo4Tiers/begin/tests/e2e/createAssignment.test.ts:14]: I also didn't know about this trick haha. So I guess Jest runner automatically calls it by just passing the function reference?

**Comment by MarioBanovac** [comment]
I like your code but I have left a few comments and would like to hear your opinion :)

To mastery!

**Review by MarioBanovac — APPROVED** [review]

File: ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart1/exercises/1_RefactoringTo4Tiers/begin/tests/e2e/assignStudentToAssignment.test.ts (hunk 1 above)
Comment (enekesabel) [reply ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart1/exercises/1_RefactoringTo4Tiers/begin/tests/e2e/assignStudentToAssignment.test.ts:18]: Neither did I, it's my first time writing BE code :D Apparently this was needed to make the tests repeatable

File: ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart1/exercises/1_RefactoringTo4Tiers/begin/tests/e2e/assignStudentToAssignment.test.ts (hunk 2 above)
Comment (enekesabel) [reply ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart1/exercises/1_RefactoringTo4Tiers/begin/tests/e2e/assignStudentToAssignment.test.ts:32]: What do you mean exactly? The usage of builders or extracting the `givenIHaveAStundentAndAnAssignment` method?

If it's the latter I stole the idea from the jest-cucumber repo. I was looking for a way to properly implement the gherkin `Background` statements, and this is the closest I got :D 

References:
https://github.com/bencompton/jest-cucumber/blob/main/examples/typescript/specs/features/backgrounds.feature
https://github.com/bencompton/jest-cucumber/blob/main/examples/typescript/specs/step-definitions/backgrounds.steps.ts

File: ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart1/exercises/1_RefactoringTo4Tiers/begin/tests/e2e/createAssignment.test.ts (hunk 3 above)
Comment (enekesabel) [reply ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart1/exercises/1_RefactoringTo4Tiers/begin/tests/e2e/createAssignment.test.ts:14]: Yes, it's basically a shorthand for 
```
beforeEach(async () => await resetDatabase())
```

File: ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart1/exercises/1_RefactoringTo4Tiers/begin/tests/e2e/assignStudentToAssignment.test.ts (hunk 2 above)
Comment (MarioBanovac) [reply ThePhasesOfCraftship/2_best_practice_first/strategicDesignPart1/exercises/1_RefactoringTo4Tiers/begin/tests/e2e/assignStudentToAssignment.test.ts:32]: I was referencing the `givenIHaveAStundentAndAnAssignment ` but now I see it from the docs that you provided that it makes sense
